nova/hypervisor.go

```diff
@@ -0,0 +1,251 @@
+package nova
+
+import (
+	"fmt"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/rackspace/gophercloud"
+	"github.com/rackspace/gophercloud/openstack/compute/v2/extensions/adminactions"
+	"github.com/rackspace/gophercloud/openstack/compute/v2/servers"
+	"github.com/rackspace/gophercloud/pagination"
+	"github.com/stackanetes/kubernetes-entrypoint/logger"
+)
+
+const (
+	retryInterval = 2
+	novaComputeBinaryName = "nova-compute"
+	enabledString = "enabled"
+	liveMigrationRetry = 3
+)
+
+// Service is a struct which represents single Openstack service
+type Service struct {
+	Status          string
+	Binary          string
+	Host            string
+	Zone            string
+	State           string
+	Disabled_reason string
+	Id              int
+}
+
+// NovaService is struct which represents Nova services returned by OpenStack API
+type NovaService struct {
+	Services []Service
+}
+
+// Node is an implementation of OpenStack hypervisor.
+type Hypervisor struct {
+	body     map[string]string
+	client   *gophercloud.ServiceClient
+	confPath string
+	hostname string
+	timeOut  time.Duration
+	vms      *[]servers.Server
+	Enabled  bool
+}
+
+
+// NovaServer is struct which represents Nova server returned by OpenStack API
+type NovaServer struct {
+	Server servers.Server
+}
+
+// New is a constructor for Hypervisor.
+func New(confPath string, timeOut int) (*Hypervisor, error) {
+	to := time.Duration(timeOut) * time.Minute
+	hostname, err := os.Hostname()
+	if err != nil {
+		return nil, fmt.Errorf("Cannot retrieve hostname: %v", err)
+	}
+	client, err := createOpenstackClient(confPath)
+	if err != nil {
+		return nil, fmt.Errorf("Cannot create openstack client: %v", err)
+	}
+	return &Hypervisor{
+		body: map[string]string{
+			"binary": "nova-compute",
+			"host": hostname,
+		},
+		client:  client,
+		confPath: confPath,
+		hostname: hostname,
+		timeOut: to,
+		Enabled: true,
+	}, nil
+}
+
+func (n *Hypervisor) novaServices() ([]Service, error) {
+	nova := new(NovaService)
+	url := n.client.ServiceURL("os-services")
+	resp, err := n.client.Request("GET", url, gophercloud.RequestOpts{
+		OkCodes: []int{200, 204},
+	})
+	if err != nil {
+		return nil, fmt.Errorf("Cannot gather openstack-service list: %v", err)
+	}
+
+	if err = getJson(resp.Body, &nova); err != nil {
+		err = fmt.Errorf("Cannot decode JSON: %v", err)
+	}
+
+	return nova.Services, err
+}
+func (n *Hypervisor) hypervisorStatus() (bool, error) {
+	services, err := n.novaServices()
+	if err != nil {
```

Comment:
n.novaServices() should be called multiple times in case of error, we can also use rety count defined as argv.